core/src/main/scala/caliban/execution/Field.scala

```diff
@@ -46,25 +46,21 @@ case class Field(
   private[caliban] val aliasedName: String =
     if (alias.isEmpty) name else alias.get
 
-  private[caliban] lazy val allFieldsUniqueNameAndCondition: Boolean = {
-    def inner(fields: List[Field]): Boolean = {
-      val headCondition = fields.head._condition
-
-      val seen = new mutable.HashSet[String]
-      seen.add(fields.head.aliasedName)
-
-      var rem = fields.tail
-      while (rem ne Nil) {
-        val f        = rem.head
-        val continue = seen.add(f.aliasedName) && f._condition == headCondition
-        if (!continue) return false
+  // TODO: Change the name to `allConditionsUnique` in the next minor version
+  private[caliban] def allFieldsUniqueNameAndCondition: Boolean =
+    fields.isEmpty || fields.tail.isEmpty || allConditionsUniqueLazy
+
+  private lazy val allConditionsUniqueLazy: Boolean = {
+    val headCondition = fields.head._condition
+    var rem           = fields.tail
+    var res           = true
+    while ((rem ne Nil) && res) {
```

Comment:
Wasn't there some optimization to assign Nil to a local variable that we used elsewhere?